File: 1248-NumberOfNiceSubArrays.py
class Solution:
    def numberOfSubarrays(self, nums: List[int], k: int) -> int:
        n = len(nums)
        def atmostK(k):
            result = 0
            odds = 0
            left = 0
            for right in range(n):
                if nums[right] % 2 != 0:
                    odds += 1
            
                while odds > k:
                    if nums[left] % 2 != 0:
                        odds -= 1
                    left += 1

                result += right - left + 1
            return result

        return atmostK(k) - atmostK(k - 1)

File: 1248-NumberOfNiceSubArrays.cpp
class Solution {
public:
    int atmostK(vector<int>& v, int k)
    {
        int left = 0, right = 0;
        int odds = 0, ans = 0;
        while (right < v.size())
        {
            if (v[right] % 2 == 1) odds++;
            
            while (odds > k) {
                if(v[left++] % 2 == 1) odds--;
            }
            ans += (right - left + 1);
            right++;
        }
        return ans;
    }
    int numberOfSubarrays(vector<int>& nums, int k) {
        return atmostK(nums, k) - atmostK(nums, k-1);
    }
};